### PR TITLE
track withdrawal proofs in relayer

### PR DESCRIPTION
**Pull request by KarishmaBothara**
closes https://github.com/cennznet/bridge-ui/issues/33

Screen Shot 2021-12-22 at 7 22 23 AM

endpoint - http://0.0.0.0:3000/withdrawals/5EYxYJVZFwa4T1nVGFadeMNWRhHPYboMdToEbiER2AzWVsLK

### CODE REVIEW

File: src/mongo/models.js [hunk 1]
```diff
@@ -20,9 +22,20 @@ const EventProcessedSchema = new Schema({
     blockHash: String,
 }, { collection: EVENT_PROCESSED });
 
+const WithdrawProofSchema = new Schema({
```
[the remark this answers is not in the document]
Comment (KarishmaBothara) [reply src/mongo/models.js:25]: updated the code as per review comment

**Review by holygits — COMMENTED** [review]
looking good, couple changes:

1) can we save the info like txHash, amount, asset id?
this will be useful for app to display.

2) return a proof by eventId: `/proofs/:proofId` so app only gets the proof it needs

3) `/withdrawals/:address`
returns the history of all withdrawals
app can extract proof id and request it if needed.
these proofs are largeish on mainnet and doesn't make sense to retrieve it all at once.
```js
{
  _id: '5xasdfasdfasdfasdsf',
   withdrawals: [
     {
        proofId: 3 // eventId,
        amount: "1000000" // wei,
        assetId:  "1001",
      },
      {
        proofId: 555,
        amount: "1000000" // wei,
        assetId:  "1",
       }
   ]
}
```

**Comment by KarishmaBothara** [comment]
> looking good, couple changes:
> 
> 1. can we save the info like txHash, amount, asset id?
>    this will be useful for app to display.
> 2. return a proof by eventId: `/proofs/:proofId` so app only gets the proof it needs
> 3. `/withdrawals/:address`
>    returns the history of all withdrawals
>    app can extract proof id and request it if needed.
>    these proofs are largeish on mainnet and doesn't make sense to retrieve it all at once.
> 
> ```js
> {
>   _id: '5xasdfasdfasdfasdsf',
>    withdrawals: [
>      {
>         proofId: 3 // eventId,
>         amount: "1000000" // wei,
>         assetId:  "1001",
>       },
>       {
>         proofId: 555,
>         amount: "1000000" // wei,
>         assetId:  "1",
>        }
>    ]
> }
> ```

Makes sense to me...

Screen Shot 2021-12-23 at 11 07 01 AM
Screen Shot 2021-12-23 at 11 07 17 AM

**Comment by eupston** [comment]
@holygits @KarishmaBothara Looks good to me, did we want to include how much longer a proof is valid (I.E how long they have to claim ETH tokens) when a user retrieves a proof via the endpoint? And/or possibly if it has already been claimed?

File: scripts/subscribeEventProof.js [hunk 2]
```diff
@@ -269,6 +272,28 @@ async function main (networkName, bridgeContractAddress) {
                 }
             })
         });
+
+    //listen to withdraw events on eth to update claimed proofs
+    pegContract.on("Withdraw", async (from, to, value, event) => {
```
Comment (KarishmaBothara) [inline scripts/subscribeEventProof.js:296]: we had observed in past that WebSocket Infura provider does not listen to event (deposit) in claim relayer.
used Alchemy provider (https://github.com/cennznet/bridge-contracts/commit/052e2f136b556816bfdfcaee4bc51e2e2900a81f)

Comment (eupston) [reply scripts/subscribeEventProof.js:296]: oo okay sweet i'll switch over to that provider

**Comment by holygits** [comment]
this PR will need some updates to track the `validators` in each proof and use new peg ABI

**Comment by KarishmaBothara** [comment]
> this PR will need some updates to track the `validators` in each proof and use new peg ABI

done

File: scripts/subscribeWithdrawTx.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        const cennznetEraTimeSeconds = 86400;
+        const cennnznetBlockTimeSec = 5;
+        let remainingSecsTillExpires = currentProofTTLEras * cennznetEraTimeSeconds;
+        //check how much time passed in current cennznet epoch
+        // const response = await axios.get(
+        //     `${process.env.UNCOVER_URI}/cennznet-explorer-api/api/scan/events?moduleId=ethBridge&eventId=AuthoritySetChange&row=100&page=0`
+        // );
+        // const uncoverEventsData = response.data;
+        // let uncoverEvents = uncoverEventsData.data;
+        // const lastEraBlockNum = uncoverEvents[uncoverEvents.length - 1].block_num;
```
Comment (eupston) [inline scripts/subscribeWithdrawTx.js]: O nice replacement could we delete this?

**Review by eupston — APPROVED** [review]

File: src/mongo/models.js [hunk 4]
```diff
@@ -6,6 +6,10 @@ const BRIDGE_CLAIM = 'claims';
 
 const EVENT_PROCESSED = 'events';
 
+const WITHDRAW_PROOF = 'withdraw_proof';
+
+const EVENT_PROOF = 'event_proof';
```
Comment (holygits) [inline src/mongo/models.js:11]: nit: convention to use camel case for these collection names
`withdrawProof` & `eventProof`

**Pull request merged**